### PR TITLE
auto-generate a somewhat reasonable content id if there isn't one

### LINKED ISSUE #2071: 2.11.22 broke some series parsing
### Expected behaviour:
In the task fill-series-list from my config, it used to fetch series from a trakt list, lookup for the tvdb name for those series, and add the names to a list

### Actual behaviour:
After 2.11.22, this doesnt work anymore, because the parsing fails for some reason, and is unable to resolve the tvdb_series_name. Either that or some execution order change...

### Steps to reproduce:
- Just run the config below or see the logs

#### Config:
```
tasks:
  fill-series-list:
    metainfo_series: yes
    thetvdb_lookup: yes
    list_clear:
      what:
        - entry_list: trakt-series
    trakt_list:
      username: "jonybat"
      list: "test"
      type: shows
    accept_all: yes
    set:
      title: "{{tvdb_series_name}}"
    list_add:
      - entry_list: trakt-series
```
  
#### Log:
2.11.21:
```
2018-01-31 21:09 VERBOSE  details       fill-series-list Produced 1 entries.
2018-01-31 21:09 DEBUG    api_tvdb      fill-series-list Looking up tvdb information for 'Westworld (2016)'. TVDB ID: 296762
2018-01-31 21:09 DEBUG    api_tvdb      fill-series-list Series <name=Westworld (2016),tvdb_id=296762> information restored from cache.
2018-01-31 21:09 VERBOSE  task          fill-series-list ACCEPTED: `Westworld` by accept_all plugin
2018-01-31 21:09 DEBUG    urlrewriter   fill-series-list Checking 1 entries
2018-01-31 21:09 VERBOSE  details       fill-series-list Summary - Accepted: 1 (Rejected: 0 Undecided: 0 Failed: 0)
2018-01-31 21:09 VERBOSE  list_add      fill-series-list adding accepted entries into entry_list - trakt-series
2018-01-31 21:09 DEBUG    api_tvdb      fill-series-list Looking up tvdb information for 'Westworld (2016)'. TVDB ID: 296762
2018-01-31 21:09 DEBUG    api_tvdb      fill-series-list Series <name=Westworld (2016),tvdb_id=296762> information restored from cache.
2018-01-31 21:09 DEBUG    api_tvdb      fill-series-list Looking up tvdb information for 'Westworld (2016)'. TVDB ID: 296762
2018-01-31 21:09 DEBUG    api_tvdb      fill-series-list Series <name=Westworld (2016),tvdb_id=296762> information restored from cache.
2018-01-31 21:09 DEBUG    entry_list    fill-series-list adding entry <Entry(title=Westworld,state=accepted)> to list trakt-series
```
2.11.22:
```
2018-01-31 21:11 VERBOSE  details       fill-series-list Produced 1 entries.
2018-01-31 21:11 DEBUG    utils.template fill-series-list Error during rendering: (UndefinedError) 'tvdb_series_name' is undefined
2018-01-31 21:11 ERROR    set           fill-series-list Could not set title for Westworld (2016): (UndefinedError) 'tvdb_series_name' is undefined
2018-01-31 21:11 DEBUG    parser_internal fill-series-list Parsing series: `Westworld (2016)` kwargs: {'identified_by': u'auto', 'name': None, 'allow_seasonless': False}
2018-01-31 21:11 DEBUG    parser_internal fill-series-list Parsing result: <SeriesParser(data=Westworld (2016),name=None,id=None,season=None,season_pack=None,episode=None,quality=None,proper=0,status=INVALID)> (in 54.946 ms)
2018-01-31 21:11 VERBOSE  task          fill-series-list ACCEPTED: `Westworld (2016)` by accept_all plugin
2018-01-31 21:11 DEBUG    urlrewriter   fill-series-list Checking 1 entries
2018-01-31 21:11 VERBOSE  details       fill-series-list Summary - Accepted: 1 (Rejected: 0 Undecided: 0 Failed: 0)
2018-01-31 21:11 VERBOSE  list_add      fill-series-list adding accepted entries into entry_list - trakt-series
2018-01-31 21:11 DEBUG    api_tvdb      fill-series-list Looking up tvdb information for 'Westworld (2016)'. TVDB ID: 296762
2018-01-31 21:11 DEBUG    api_tvdb      fill-series-list Series <name=Westworld (2016),tvdb_id=296762> information restored from cache.
2018-01-31 21:11 DEBUG    api_tvdb      fill-series-list Looking up tvdb information for 'Westworld (2016)'. TVDB ID: 296762
2018-01-31 21:11 DEBUG    api_tvdb      fill-series-list Series <name=Westworld (2016),tvdb_id=296762> information restored from cache.
2018-01-31 21:11 DEBUG    api_tvdb      fill-series-list Looking up tvdb information for 'Westworld (2016)'. TVDB ID: 296762
2018-01-31 21:11 DEBUG    api_tvdb      fill-series-list Series <name=Westworld (2016),tvdb_id=296762> information restored from cache.
2018-01-31 21:11 DEBUG    entry_list    fill-series-list adding entry <Entry(title=Westworld (2016),state=accepted)> to list trakt-series
```
### Additional information:

- FlexGet version: 2.11.22 and above
- Python version: Tested in 2.7.14 and 3.5.2
- Installation method: pip
- Using daemon (yes/no): Either
- OS and version: Fedora 25 and Ubuntu 16.04

### PR DESCRIPTION
**Pull request by cvium**
### Motivation for changes:
None
### Detailed changes:
- Automatically generate a content id based on entry type if the entry hasn't been parsed by internal or guessit

### Addressed issues:
- Fixes #2071 

#### To Do:

- [x] Test

### CODE REVIEW

**Comment by gazpachoking** [comment]
I'm 👍 on reverting metainfo movie/series to not use the 'id' field again.
My other feelings:
- I don't really like the field name 'id' at all, seems too generic for what it is. Perhaps 'content_id' or 'media_id' or something.
- Maybe that field should be populated by a metainfo_media_id plugin or something rather than have individual parser type plugins populate it. This plugin would register it as a lazy field, then if the required fields were present (series_name&series_id/movie_name) generate it. This plugin could either be built in, or automatically invoked by plugins that need it (timeframe/upgrade.)

**Comment by stevezau** [comment]
I just noticed this.. What's the idea behind this revert? I'm asking as this will break my config.

**Comment by cvium** [comment]
The parsers need to set the id field since it's set based on the parsed identifiers etc. We can rename it to media_id before setting it on the entry, but it's not really plausible as-is to create a separate plugin that replaces that part in the parsers.

**Comment by gazpachoking** [comment]
> The parsers need to set the id field since it's set based on the parsed identifiers etc. We can rename it to media_id before setting it on the entry, but it's not really plausible as-is to create a separate plugin that replaces that part in the parsers.

My idea was that since the media ids are based on other fields, it could be a lazy field, which just checks the requisite fields are there.

**Comment by AnthonyGuerreiro** [comment]
Tested against the [original issue](https://github.com/Flexget/Flexget/issues/2071) and opened another [PR](https://github.com/Flexget/Flexget/pull/2298) on top of this one.

**Comment by gazpachoking** [comment]
I think I'm liking this better now. A couple more notes though:
- I think utils/entry.py with the is_show stuff isn't so great, we can move those straight to the media_id plugin.
- I don't think it should generate a media_id for a 'show', only for an episode. (I think that would also resolve the issue that #2298 is trying to solve.)
- I think we should eliminate the 'id' field altogether. It was introduced for the timeframe and upgrade plugin, and media_id is now taking that role. i.e. this https://github.com/Flexget/Flexget/blob/0f7f805a43e25e27fce195b91228f911bf4c6b1e/flexget/plugins/filter/series.py#L1314-L1315 I can't seem to find where that field was even being filled for movies...
The check for that field in the media_id plugin could also be removed.

**Comment by FallingReign** [comment]
do we have an ETA on this merge?

**Comment by zilexa** [comment]
Could I help test this?

**Comment by marine1988** [comment]
is there any news?
Is this fixed?

**Comment by Jonybat** [comment]
So, i'd like to update, but i'm stuck on a version from 2 years ago. How can i help with this?

File: flexget/plugins/metainfo/media_id.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        if config is False:
+            return
+        for entry in task.entries:
+            entry.register_lazy_func(self.get_media_id, ['media_id'])
+
+    @staticmethod
+    def get_media_id(entry):
+        # Try to generate a media id based on available parser fields
+        media_id = None
+        if bool(entry.get('movie_name')):
```
Comment (gazpachoking) [inline flexget/plugins/metainfo/media_id.py]: These `bool` casts are unnecessary here.

**Comment by Jonybat** [comment]
Thank you @cvium

**Pull request merged**